av.c

```diff
@@ -521,31 +524,34 @@ Perl_newAVhv(pTHX_ HV *ohv)
 
     bool tied = SvRMAGICAL(ohv) && mg_find(MUTABLE_SV(ohv), PERL_MAGIC_tied);
 
-    AV *ret = newAV();
+    U32 nkeys = hv_iterinit(ohv);
+    /* This number isn't perfect but it doesn't matter; it only has to be
+     * close to make the initial allocation about the right size
+     */
+    AV *ret = newAV_alloc_xz( nkeys ? nkeys * 2 : 2);
 
     /* avoid ret being leaked if croak when calling magic below */
     EXTEND_MORTAL(1);
     PL_tmps_stack[++PL_tmps_ix] = (SV *)ret;
     SSize_t ret_at_tmps_ix = PL_tmps_ix;
 
-    U32 nkeys = hv_iterinit(ohv);
-    /* This number isn't perfect but it doesn't matter; it only has to be
-     * close to make the initial allocation about the right size
-     */
-
-    av_extend(ret, nkeys * 2);
 
     HE *he;
+    SSize_t i = 0;
     while((he = hv_iternext(ohv))) {
+        if (i > AvMAX(ret))
+            av_extend(ret, i);
+
         if(tied) {
-            av_push(ret, newSVsv(hv_iterkeysv(he)));
-            av_push(ret, newSVsv(hv_iterval(ohv, he)));
+            AvARRAY(ret)[i++] = newSVsv(hv_iterkeysv(he));
+            AvARRAY(ret)[i++] = newSVsv(hv_iterval(ohv, he));
         }
         else {
-            av_push(ret, newSVhek(HeKEY_hek(he)));
-            av_push(ret, HeVAL(he) ? newSVsv(HeVAL(he)) : &PL_sv_undef);
+            AvARRAY(ret)[i++] = newSVhek(HeKEY_hek(he));
+            AvARRAY(ret)[i++] = HeVAL(he) ? newSVsv(HeVAL(he)) : &PL_sv_undef;
         }
     }
+    AvFILLp(ret) = i - 1;
```

Comment:
note that this could lead to set it to `-1`?

Reply:
probably not, we abort earlier after checking the number of entries, if it s `= 0`

Reply:
Yeah, in `newAVav()`, that assignment shouldn't be reached if `count == 0`. I think it could be reached in the `newAVhv()` case if there were actually no keys in `ohv`, but `AvFILLp(ret)` really should `== 1` if that's the case.